File: gov/auth/peer_t.cpp
#include "peer_t.h"
#include <us/gov/engine/daemon.h>

using namespace us::gov::auth;
using namespace std;

constexpr array<const char*,peer_t::num_stages> peer_t::m_stagestr;

void peer_t::verification_completed() {
    if (unlikely(!verification_is_fine()) || unlikely(!authorize(get_pubkey()))) {
        return;
    }
    m_stage=authorized;
}
void peer_t::dump(ostream& os) const {
    os << "auth: " << this << ' ' << get_pubkey() << " " << m_stagestr[m_stage] << endl;
}

File: sdk/java/us/gov/auth/peer_t.java
package us.gov.auth;
import java.net.Socket;
import java.security.PublicKey;

public class peer_t extends us.gov.id.peer_t {
        public enum stage_t {
            denied,
            authorized,
            num_stages;

            public short asShort() {
                return (short)ordinal();
            }
            public static stage_t fromShort(short i) {
                return stage_t.values()[i];
            }
        };

        public String[] stagestr = {"denied","authorized"};

        public peer_t(Socket sock) {
            super(sock);
        }

        @Override
       public void verification_completed() {
            //super.verification_completed();
            if (!verification_is_fine()) {
                return;
            }
           	if (!authorize(pubkey)) {
                return;
           	}
           	stage=stage_t.authorized;
        }

    	public boolean authorize(PublicKey pub) {
            return false;
        }
        //virtual string run_auth() override;

        public stage_t stage=stage_t.denied;
}
